src/ostrowski/rationals/bounded.lean

```diff
@@ -0,0 +1,364 @@
+    have p: padic_val_rat p q = 0,
+    {
+      unfold padic_val_rat,
+      simp [q_prime.1, prime.ne_one p_prime],
+      suffices h: p^0 ∣ q ∧ ¬ p^1 ∣ q,
+      {
+        norm_cast,
+        rw roption.get_eq_iff_eq_some,
+        rw multiplicity.eq_some_iff.2 h,
+        refl,
+      },
+      split,
+      {
+        simp,
+      },
+      {
+        by_contra p_div_q,
+        apply h,
+        rw ← associated_iff_eq,
+        simp at p_div_q,
+        have q_div_p := dvd_symm_of_irreducible
+          (irreducible_of_prime p_prime)
+          (irreducible_of_prime q_prime)
+          p_div_q,
+        exact associated_of_dvd_dvd p_div_q q_div_p,
+      },
+    },
+    unfold padic_norm,
+    rw p,
+    simp [q_prime.1],
+  },
+end
+
+lemma abs_val_bounded_q (abv : ℚ → ℝ) [habv : is_absolute_value abv]
+  (all_nat_le_one: (∀ z : ℕ, abv z ≤ 1))
+  (p: ℕ) (p_prime: prime p) (abvp_lt_one: abv p < 1):
+  ∃ α: ℝ, 0 < α ∧ ∀ q: ℕ, prime q →
+    (p = q → abv q = (1/p) ^ α) ∧ (p ≠ q → abv q = 1 ^ α) :=
+begin
+  -- We set `α` such that it Just Works™
+  set α := - real.log (abv p) / real.log p with α_def,
+  have zero_lt_α: 0 < α,
+  {
+    apply div_pos,
+    have one_lt_abvp: abv p < 1,
+    {
+      rw ← gt_iff_lt,
+      exact abvp_lt_one,
+    },
+    apply neg_pos_of_neg,
+    have p_ne_zero_ℚ: (p: ℚ) ≠ 0,
+    { norm_cast, linarith only [nat.prime.pos $ nat.prime_iff_prime.2 p_prime], },
+    exact (real.log_neg_iff ((abv_pos abv).2 p_ne_zero_ℚ)).2
+      one_lt_abvp,
+    have p_pos_ℝ: 0 < (p: ℝ),
+    { norm_cast, exact (nat.prime.pos $ nat.prime_iff_prime.2 p_prime), },
+    exact (real.log_pos_iff p_pos_ℝ).2
+        (by { norm_cast, exact (nat.prime.one_lt $ nat.prime_iff_prime.2 p_prime), }),
+  },
+  use [α, zero_lt_α],
+
+  intros q q_prime,
+
+  split,
+  intro h,
+  {
+    -- When `p = q`, we just need to show that `abv p = p ^ (- α)`.
+    have zero_lt_p: 0 < (p: ℝ),
+    { norm_cast, exact nat.lt_of_le_and_ne (zero_le p) (ne.symm (p_prime.1)), },
+    -- The result is clear by definition of `α`.
+    -- The calculus that leads to it is yet long to formalize...
+    rw ← h,
+    suffices h: real.log (abv p) = real.log ((1/p) ^ α),
+    {
+      have p₁: abv p > 0,
+      {
+        apply gt_iff_lt.2,
+        apply (abv_pos abv).2,
+        norm_cast,
+        exact p_prime.1,
+      },
+      have p₂: (1/p: ℝ)^α > 0,
+      {
+        apply real.rpow_pos_of_pos,
+        rw one_div_pos,
+        -- norm_cast,
+        exact zero_lt_p,
+      },
+      apply le_antisymm,
+      rw ← real.log_le_log p₁ p₂,
+      exact le_of_eq h,
+      rw ← real.log_le_log p₂ p₁,
+      exact le_of_eq (eq.symm h),
+    },
+    have one_lt_p: 1 < p,
+    from (nat.prime.one_lt ∘ nat.prime_iff_prime.2) p_prime,
+    calc real.log (abv p) = real.log (abv p) * 1
+        : by ring
+      ... = real.log (abv p) * (real.log p / real.log p)
+        : by { rw div_self, symmetry, apply ne_of_lt ∘ real.log_pos,
```

Comment:
symmetry, apply ne_of_lt circ real.log_pos is exactly apply eq.symm (ne_of_lt circ real.log_pos) I believe
not necessarily something good to change, it might not be worth the effort

it would be better though if you could factor out the requirements of each rhs and put them in a have before the calc phase which is a bit hardcore of course

therefore, the rhs would be deadsimple applications and minimal tactics

Reply:
Done.